src/platform/ESP32/NetworkCommissioningDriver.h

```diff
@@ -0,0 +1,129 @@
+        item.channel  = mpscanResults[iternum].primary;
+        item.wiFiBand = chip::DeviceLayer::NetworkCommissioning::WiFiBand::k2g4;
+        item.rssi     = mpscanResults[iternum].rssi;
+        memcpy(item.ssid, mpscanResults[iternum].ssid, item.ssidLen);
+        memcpy(item.bssid, mpscanResults[iternum].bssid, 6);
+
+        iternum++;
+        return true;
+    }
+    void Release() override {}
+
+private:
+    size_t mSize;
+    wifi_ap_record_t * mpscanResults;
+    uint8_t iternum = 0;
+};
+
+class ESPWiFiDriver final : public WiFiDriver
+{
+public:
+    class WiFiNetworkIterator final : public NetworkIterator
+    {
+    public:
+        WiFiNetworkIterator(ESPWiFiDriver * aDriver) : driver(aDriver) {}
+        size_t Count() override;
+        bool Next(Network & item) override;
+        void Release() override { delete this; }
+        ~WiFiNetworkIterator() = default;
+
+    private:
+        ESPWiFiDriver * driver;
+        bool exhausted = false;
+    };
+
+    struct WiFiNetwork
+    {
+        char ssid[DeviceLayer::Internal::kMaxWiFiSSIDLength];
+        uint8_t ssidLen = 0;
+        char credentials[DeviceLayer::Internal::kMaxWiFiKeyLength];
+        uint8_t credentialsLen = 0;
+    };
+
+    // BaseDriver
+    NetworkIterator * GetNetworks() override { return new WiFiNetworkIterator(this); }
+    CHIP_ERROR Init() override;
+    CHIP_ERROR Shutdown() override;
+
+    // WirelessDriver
+    uint8_t GetMaxNetworks() override { return 1; }
+    uint8_t GetScanNetworkTimeoutSeconds() override { return 10; }
+    uint8_t GetConnectNetworkTimeoutSeconds() override { return 20; }
+
+    CHIP_ERROR CommitConfiguration() override;
+    CHIP_ERROR RevertConfiguration() override;
+
+    Status RemoveNetwork(ByteSpan networkId) override;
+    Status ReorderNetwork(ByteSpan networkId, uint8_t index) override;
+    void ConnectNetwork(ByteSpan networkId, ConnectCallback * callback) override;
+
+    // WiFiDriver
+    Status AddOrUpdateNetwork(ByteSpan ssid, ByteSpan credentials) override;
+    void ScanNetworks(ByteSpan ssid, ScanCallback * callback) override;
+
+    CHIP_ERROR ConnectWiFiNetwork(const char * ssid, uint8_t ssidLen, const char * key, uint8_t keyLen);
+    void OnConnectWiFiNetwork();
+    void OnScanWiFiNetworkDone();
+    static ESPWiFiDriver & GetInstance()
+    {
+        static ESPWiFiDriver instance;
+        return instance;
+    }
+
+private:
+    bool NetworkMatch(const WiFiNetwork & network, ByteSpan networkId);
+    CHIP_ERROR StartScanWiFiNetworks(ByteSpan ssid);
+
+    WiFiNetworkIterator mWiFiIterator = WiFiNetworkIterator(this);
+    WiFiNetwork mSavedNetwork;
+    WiFiNetwork mStagingNetwork;
+    ScanCallback * mpScanCallback;
+    ConnectCallback * mpConnectCallback;
+};
+#endif // CHIP_DEVICE_CONFIG_ENABLE_WIFI
+
+#if CHIP_DEVICE_CONFIG_ENABLE_THREAD
+// TODO: Add Thread Driver for ESP32 platform
```

Comment:
is ESP32 expected to ever support thread? Otherwise for now we could just add a `#error` instead.

Reply:
The Thread driver will be added later for ESP32H2